cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -316,38 +335,62 @@ private void processMetrics() {
 
         if (currentMessageId != null) {
           topicMessageIds.put(topicIdMetaKey, currentMessageId);
+        } else {
+          return metricsProcessIntervalMillis; // sleep for the max amount of time allowed since no message is fetched
```

Comment:
I think you still need to call `tryPersist` even though there is no message fetched for this partition. Otherwise, you may be delaying persisting metrics. E.g. considering you have two threads, the following might happen:

1. Thread 1 fetched some metrics
2. Thread 1 call `tryPersist` and successfully acquired the `persistingFlag`.
3. Thread 1 copied the records
4. Thread 2 fetched some metrics
5. Thread 2 call `tryPersist`. Since thread is still holding the flag, it will just return
6. Thread 1 persisted the metrics and release the flag.
7. Both Thread 1 and Thread 2 have empty fetch.

You can see that metrics fetched by thread 2 in step 4 will not get persisted with your updated logic.